### ISSUE TITLE
SdFat examples do not compile with Apollo3 Core v2.x 

### ISSUE THREAD

**Issue by adamgarbo** [sparkfun/Arduino_Apollo3#370]
Hi there,

I discovered today that SdFat (v2.0.5) library examples are not able to compile when using Apollo3 Core v2.0.5 (or v2.0.6). The error message received is included below.

From a quick search on the issue, it seems like it's related to the Mbed OS SPI library and is only affecting SdFat. The basic SD library is still able to compile correctly.

Cheers,
Adam

```
/Users/adam/Documents/Arduino/libraries/SdFat/src/SpiDriver/SdSpiArtemis.cpp: In member function 'uint8_t SdSpiArduinoDriver::receive(uint8_t*, size_t)':
/Users/adam/Documents/Arduino/libraries/SdFat/src/SpiDriver/SdSpiArtemis.cpp:50:10: error: 'arduino::SPIClass' {aka 'class arduino::HardwareSPI'} has no member named 'transferIn'; did you mean 'transfer'?
   m_spi->transferIn(buf, count);
          ^~~~~~~~~~
          transfer
/Users/adam/Library/Arduino15/packages/SparkFun/tools/arm-none-eabi-gcc/8-2018-q4-major/bin/arm-none-eabi-g++ -include /Users/adam/Library/Arduino15/packages/SparkFun/hardware/apollo3/2.0.4/variants/SFE_ARTEMIS_MM_PB/mbed/mbed_config.h -include /Users/adam/Library/Arduino15/packages/SparkFun/hardware/apollo3/2.0.4/cores/arduino/sdk/ArduinoSDK.h -iprefix /Users/adam/Library/Arduino15/packages/SparkFun/hardware/apollo3/2.0.4/cores/ @/Users/adam/Library/Arduino15/packages/SparkFun/hardware/apollo3/2.0.4/variants/SFE_ARTEMIS_MM_PB/mbed/.cxx-flags -MMD -DARDUINO=10813 -DARDUINO_APOLLO3_SFE_ARTEMIS_MM_PB -DARDUINO_ARCH_MBED -DARDUINO_ARCH_APOLLO3 -DMBED_NO_GLOBAL_USING_DIRECTIVE -DCORDIO_ZERO_COPY_HCI @/Users/adam/Library/Arduino15/packages/SparkFun/hardware/apollo3/2.0.4/variants/SFE_ARTEMIS_MM_PB/mbed/.cxx-symbols -I/Users/adam/Library/Arduino15/packages/SparkFun/hardware/apollo3/2.0.4/cores/arduino -I/Users/adam/Library/Arduino15/packages/SparkFun/hardware/apollo3/2.0.4/variants/SFE_ARTEMIS_MM_PB -I/Users/adam/Documents/Arduino/libraries/SdFat/src -I/Users/adam/Library/Arduino15/packages/SparkFun/hardware/apollo3/2.0.4/libraries/SPI/src -I/Users/adam/Library/Arduino15/packages/SparkFun/hardware/apollo3/2.0.4/cores/arduino -I/Users/adam/Library/Arduino15/packages/SparkFun/hardware/apollo3/2.0.4/cores/arduino/mbed-bridge -I/Users/adam/Library/Arduino15/packages/SparkFun/hardware/apollo3/2.0.4/cores/arduino/mbed-bridge/core-api -I/Users/adam/Library/Arduino15/packages/SparkFun/hardware/apollo3/2.0.4/cores/arduino/mbed-bridge/core-api/api/deprecated @/Users/adam/Library/Arduino15/packages/SparkFun/hardware/apollo3/2.0.4/variants/SFE_ARTEMIS_MM_PB/mbed/.includes /Users/adam/Documents/Arduino/libraries/SdFat/src/SpiDriver/SdSpiParticle.cpp -o /var/folders/v3/5019zqsn647bxggzbh29j9mh0000gn/T/arduino_build_486459/libraries/SdFat/SpiDriver/SdSpiParticle.cpp.o
Using library SdFat at version 2.0.5 in folder: /Users/adam/Documents/Arduino/libraries/SdFat 
Using library SPI at version 2.0.0 in folder: /Users/adam/Library/Arduino15/packages/SparkFun/hardware/apollo3/2.0.4/libraries/SPI 
exit status 1
Error compiling for board Artemis MicroMod Processor.
```

**Comment by adamgarbo** [comment]
An update that while there are some small issues in the SdFat that were preventing the code from compiling (see: https://github.com/greiman/SdFat/issues/280), once these are addressed, it appears that there are Mbed OS-related issues that manifest themselves once the code is uploaded:

```
#include <SdFat.h>
#include <SPI.h>

SdFs sd;

void setup() {
  SPI.begin();
  Serial.begin(115200); 
  while (!Serial);
  
  if (!sd.begin(SdSpiConfig(41, DEDICATED_SPI)))
    Serial.println("Warning: microSD initialization failed.");
}

void loop() {}
```

Output: 

```
11:15:25.929 -> ++ MbedOS Error Info ++
11:15:25.929 -> Error Status: 0x80FF0144 Code: 324 Module: 255
11:15:25.929 -> Error Message: Assertion failed: AM_HAL_STATUS_SUCCESS == am_hal_iom_configure(obj->iom.handle, &obj->iom.cfg)
11:15:25.929 -> Location: 0x22BA1
11:15:25.929 -> File: mbed-os/targets/TARGET_Ambiq_Micro/TARGET_Apollo3/device/iom_api.c+35
11:15:25.929 -> Error Value: 0x0
11:15:25.929 -> Current Thread: main Id: 0x10004650 Entry: 0x23B41 StackSize: 0x1000 StackMem: 0x10006298 SP: 0x10007154 
11:15:25.962 -> For more info, visit: https://mbed.com/s/error?error=0x80FF0144&tgt=SFE_ARTEMIS_MM_PB
11:15:25.962 -> -- MbedOS Error Info --
```

**Comment by adamgarbo** [comment]
To update this issue, I believe I may have found the cause of the Mbed OS error.

It appears that this error will only manifest itself when attempting to initialize the microSD at SPI speeds of 50 MHz. From my quick tests, speeds of 4, 8, and 16 MHz work correctly. I am assuming that SdFat defines `DEDICATED_SPI` as 50 MHz somewhere in the library.

The following code does ***not*** work:

```
if (!sd.begin(CS, DEDICATED_SPI))
{
  Serial.println("Warning: microSD initialization failed.");
}

if (!sd.begin(CS, SD_SCK_MHZ(50)))
{
  Serial.println("Warning: microSD initialization failed.");
}
```

The following code ***does*** work:
```
if (!sd.begin(CS, SD_SCK_MHZ(4))) // Or 8 or 16 MHz
{
  Serial.println("Warning: microSD initialization failed.");
}
```

While version 1.2.1 of the Core doesn't encounter any issues with `DEDICATED_SPI` or 50 MHz, I have experienced the same issue with SAMD-based processors. I suspect in this case it may be related to Mbed OS SPI. I'll leave it up to you @Wenn0101 if you'd like to close this issue for now and just have a warning that 50 MHz isn't supported. I'm sure you must feel like you're playing whack-a-mole with all the bugs for v2.x.

Cheers,
Adam

**Comment by nseidle** [comment]
It's a 48MHz part so 50MHz SPI is not possible.

We had an issue in v1 core where it was [not correctly setting the IOM to 48MHz](https://github.com/sparkfun/Arduino_Apollo3/pull/89/files#diff-cab308c668f25b7d3e38a7e32fe603eeR126), so I'm not sure if that got integrated into v2.

**Comment by Wenn0101** [comment]
Examining the error that mbed is throwing, it is actually an error that we put in our spi driver to error out if we fail to properly config. We can see this in the `Assertion failed:  AM_HAL_STATUS_SUCCESS == am_hal_iom_configure(obj->iom.handle, &obj->iom.cfg)`. We are not getting a success from the hal layer configuration function. Looking at this function, trying to set a frequency over 48MHz will indeed cause this to happen.

```
// Max Frequency supported in HAL
#define AM_HAL_IOM_MAX_FREQ AM_HAL_IOM_48MHZ
...
if (psConfig->ui32ClockFreq > AM_HAL_IOM_MAX_FREQ)
{
    return AM_HAL_STATUS_INVALID_ARG;
}
```

**Comment by adamgarbo** [comment]
Hi folks,

The 48 MHz limit makes complete sense. Bill fixed the compilation errors that I encountered earlier in v2.0.6 of SdFat: https://github.com/greiman/SdFat/releases/tag/2.0.6

I haven't actually been able to find how the speed of DEDICATED_SPI is defined yet. It a new way of initializing the microSD in SdFat v2.x.

```
// Select the fastest interface. Assumes no other SPI devices.
#if ENABLE_DEDICATED_SPI
#define SD_CONFIG SdSpiConfig(SD_CS_PIN, DEDICATED_SPI)
#else  // ENABLE_DEDICATED_SPI
#define SD_CONFIG SdSpiConfig(SD_CS_PIN, SHARED_SPI)
#endif  // ENABLE_DEDICATED_SPI

  // Initialize SD.
  if (!sd.begin(SD_CONFIG)) {
    sd.initErrorHalt(&Serial);
  }
```

**Comment by adamgarbo** [comment]
Hey @Wenn0101,

I confirmed with Bill that the library is setting it to 50 MHz. See: https://github.com/greiman/SdFat/issues/282

So it sounds like the Apollo3 Core will need to do the limiting to 48 MHz?

Cheers,
Adam

**Comment by Wenn0101** [comment]
@adamgarbo My apologies. For some reason I had simplified this in my head to a "Do Not Fix" that I thought I just needed to close out. And that this was a problem of it asking for something we weren't capable of.

I missed your second reply referencing the issue you made in the SdFat library.
There is a bit of a coding philosophical question on whether it is best to error out or set to the closes bound when asked to set something out of bounds. It seems that we chose the opposite answer to this question as is typical for arduino. It should be pretty easy to change this in our SPI library, sorry that I missed that there was still an action needed here and didn't get it in the release that just went out.

**Comment by adamgarbo** [comment]
Hi @Wenn0101,

Thanks for the update. As long as users explicitly set the speed to 48 MHz or below and don't attempt to use DEDICATED_SPI or SPI_FULL_SPEED (i.e. 50 MHz) there shouldn't be any issues. I think it makes sense to have the SPI library set the speed to the max speed when provided with an out-of-bounds value, which I believe is the approach of v1 of the Apollo3 Core.

Cheers,
Adam

**Comment by Wenn0101** [comment]
This was fixed in v2.1.1
